clients/go/admin/deviceflow/token_orchestrator.go

```diff
@@ -0,0 +1,236 @@
+	ts := t.clientConfig.TokenSource(ctx, token)
+	var refreshedToken *oauth2.Token
+	var err error
+	refreshedToken, err = ts.Token()
+	if err != nil {
+		logger.Warnf(ctx, "failed to refresh the token due to %v and will be doing re-auth", err)
+		return nil, err
+	}
+
+	if refreshedToken != nil {
+		logger.Debugf(ctx, "got a response from the refresh grant for old expiry %v with new expiry %v",
+			token.Expiry, refreshedToken.Expiry)
+		if refreshedToken.AccessToken != token.AccessToken {
+			if err = t.tokenCache.SaveToken(refreshedToken); err != nil {
+				logger.Errorf(ctx, "unable to save the new token due to %v", err)
+				return nil, err
+			}
+		}
+	}
+
+	return refreshedToken, nil
+}
+
+// FetchTokenFromCacheOrRefreshIt fetches the token from cache and refreshes it if it'll expire within the
+// Config.TokenRefreshGracePeriod period.
+func (t TokenOrchestrator) FetchTokenFromCacheOrRefreshIt(ctx context.Context) (token *oauth2.Token, err error) {
+	token, err = t.tokenCache.GetToken()
+	if err != nil {
+		return nil, err
+	}
+
+	if !token.Valid() {
+		return nil, fmt.Errorf("token from cache is invalid")
+	}
+
+	// If token doesn't need to be refreshed, return it.
+	if time.Now().Before(token.Expiry.Add(-t.cfg.TokenRefreshGracePeriod.Duration)) {
+		logger.Infof(ctx, "found the token in the cache")
+		return token, nil
+	}
+	token.Expiry = token.Expiry.Add(-t.cfg.TokenRefreshGracePeriod.Duration)
+
+	token, err = t.RefreshToken(ctx, token)
+	if err != nil {
+		return nil, fmt.Errorf("failed to refresh token using cached token. Error: %w", err)
+	}
+
+	if !token.Valid() {
+		return nil, fmt.Errorf("refreshed token is invalid")
+	}
+
+	err = t.tokenCache.SaveToken(token)
+	if err != nil {
+		return nil, fmt.Errorf("failed to save token in the token cache. Error: %w", err)
+	}
+
+	return token, nil
+}
+
+// StartDeviceAuthorization will initiate the OAuth2 device authorization flow.
+func (t TokenOrchestrator) StartDeviceAuthorization(ctx context.Context, dareq DeviceAuthorizationRequest) (*DeviceAuthorizationResponse, error) {
+	v := url.Values{cliendID: {dareq.ClientID}, scope: {dareq.Scope}}
+	httpReq, err := http.NewRequest("POST", t.clientConfig.DeviceEndpoint, strings.NewReader(v.Encode()))
+	if err != nil {
+		return nil, err
+	}
+	httpReq.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+
+	httpResp, err := ctxhttp.Do(ctx, nil, httpReq)
+	if err != nil {
+		return nil, err
+	}
+
+	body, err := io.ReadAll(httpResp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("device authorization request failed due to  %v", err)
+	}
+	if httpResp.StatusCode != http.StatusOK {
+		return nil, &oauth2.RetrieveError{
+			Response: httpResp,
+			Body:     body,
+		}
+	}
+
+	daresp := &DeviceAuthorizationResponse{}
+	err = json.Unmarshal(body, &daresp)
+	if err != nil {
+		return nil, err
+	}
+
+	fmt.Printf("Please open the browser at the url %v and enter following verification code %v\nOR\n", daresp.VerificationURL, daresp.UserCode)
```

Comment:
The example you showed me with Google IdP actually has the user code as part of the url... is that not a supported mechanism by all IdPs?